### PR TITLE
Sign up form

### LINKED ISSUE #28: Sign up and Sign in page
Build sign up and sign in pages, create snapshot and unit test. Connect it to Firebase Auth. Add ability to sign in and sign up using Facebook and Google. Add ability to reset password.

### PR DESCRIPTION
**Pull request by Jamil-Bailony**
- [x] build the sign up form

- [x] add the styles

- [x] add the required validations for inputs

- [x] build unite tests

## Screenshots

image

image

image

image

Closes #28

### CODE REVIEW

File: src/components/SignUp/__snapshots__/index.test.js.snap [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+              <a
+                href="/terms"
+              >
+                Terms of Service
+              </a>
+               and our 
+              <a
+                href="/policy"
+              >
+                Privecy Policy
```
Comment (louisrli) [inline src/components/SignUp/__snapshots__/index.test.js.snap]: privacy

File: src/components/SignUp/__snapshots__/index.test.js.snap [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+          <button
+            className="submitBtn"
+            type="submit"
+          >
+            Sign in
+          </button>
+        </form>
+      </div>
+      <p>
+        Alredy have an account?
```
Comment (louisrli) [inline src/components/SignUp/__snapshots__/index.test.js.snap]: already

File: src/components/SignUp/__snapshots__/index.test.js.snap [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            <label
+              htmlFor="acceptTerms"
+            >
+              <input
+                id="acceptTerms"
+                name="acceptTerms"
+                type="checkbox"
+                value="accepted"
+              />
+              I have read and agree with
```
Comment (louisrli) [inline src/components/SignUp/__snapshots__/index.test.js.snap]: agree with the

File: src/components/SignUp/index.jsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import twitterIcon from "../../images/icons/twitterIcon.svg";
+import facebookIcon from "../../images/icons/facebookIcon.svg";
+import "./index.scss";
+
+export default function SignUp() {
+  return (
+    <Container fluid="md" className="signUpSection">
+      <Row>
+        <Col className="signInImg" md={6} xs={12}>
+          <img src={womenSupportEachOther} alt="Women Group Hugging Together" />
```
Comment (louisrli) [inline src/components/SignUp/index.jsx]: fix capitalization

**Review by louisrli — COMMENTED** [review]
split the pr, won't review the second half until its split

**Comment by Jamil-Bailony** [comment]
> split the pr, won't review the second half until its split

It has been spilt. Sorry I though that you accepted and merged the [sign up PR](https://github.com/ReCoded-Org/capstone-istanbul-safe_place/pull/62). I will wait until you approve and merge that PR, then I will update this one.

@louisrli

**Comment by louisrli** [comment]
if you want reviews ahead of time you remove the files that are not affected by this PR and isolate it, even if the tests dont pass. generally speaking reviews shouldnt block the ability for other PRs to get reviews

in some UIs (such as paid github) you can specify an underlying "diff" branch so that this is not a problem, but we dont have paid github so you just have to remove the files

**Comment by Jamil-Bailony** [comment]
You can look at it right now, there are just 5 files @louisrli

File: src/components/SignUpForm/__snapshots__/index.test.js.snap [hunk 5]
```diff
@@ -0,0 +1,102 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (louisrli) [inline src/components/SignUpForm/__snapshots__/index.test.js.snap]: you don't need both these snapshots as they're basically the same

File: src/components/SignUpForm/index.jsx [hunk 6]
```diff
@@ -1,5 +1,150 @@
...
+const schema = Joi.object({
+  firstName: Joi.string().required().messages({
+    "string.empty": `Please provide your first name`,
+  }),
+  lastName: Joi.string().required().messages({
+    "string.empty": `Please provide your last name`,
+  }),
+  email: Joi.string()
+    .required()
+    .email({ tlds: { allow: ["com", "net", "edu"] } })
```
Comment (louisrli) [inline src/components/SignUpForm/index.jsx]: delete the tlds. no reason to restrict what emails can register

File: src/components/SignUpForm/index.jsx [hunk 7]
```diff
@@ -1,5 +1,150 @@
...
+            <input
+              className={errors.password && "inputError"}
+              type="password"
+              placeholder="Password"
+              name="password"
+              aria-label="password"
+              ref={register()}
+            />
+            {errors?.password && (
+              <InputErrorMessage message={errors.password.message} />
```
Comment (louisrli) [inline src/components/SignUpForm/index.jsx]: all these can be folded into a common fn or component that takes errors?.x and if it's non null it renders errors?.x.messages. then reused throughout all your components.

try to find where you are copy pasting this pattern and fix

File: src/components/SignUpForm/index.jsx [hunk 8]
```diff
@@ -1,5 +1,150 @@
...
+              className={errors.acceptTerms && "inputError"}
+              type="checkbox"
+              id="acceptTerms"
+              name="acceptTerms"
+              value="accepted"
+              ref={register()}
+            />
+            I have read and agree with the&nbsp;
+            <a href="/terms">Terms of Service</a>
+            &nbsp;and our&nbsp;
```
Comment (louisrli) [inline src/components/SignUpForm/index.jsx]: actually you want {' '} here not nbsp because you want these to be breakable on spaces

File: src/components/SignUpForm/index.test.js [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+    );
+    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe("test");
+    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
+    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
+      "password"
+    );
+    expect(screen.getByRole("checkbox").checked).toBe(true);
+  });
+
+  it("should display matching error when email domain is invalid", async () => {
```
Comment (louisrli) [inline src/components/SignUpForm/index.test.js]: delete this feature. in genera you don't want to make assumptions about users emails, names, etc. you will block users randomly as a result

File: src/components/SignUpForm/index.test.js [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+        value: "password",
+      },
+    });
+
+    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
+      target: {
+        value: "password",
+      },
+    });
+
```
Comment (louisrli) [inline src/components/SignUpForm/index.test.js]: these tests need more comments. I'd say one for each user action you're mimicking. also this usage of getbyrole checkbox is bad testing practice because you are assuming there is only one checkbox. if your teammate adds another one, such as a checkbox to sign up for a newsletter, these tests will break. 

this applies for all getbyrole or similar calls where you are currently relying on the fact that there is only one of them (if any)

File: src/components/SignUpForm/index.test.js [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
+      target: {
+        value: "test@test.co",
+      },
+    });
+
+    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
```
Comment (louisrli) [inline src/components/SignUpForm/index.test.js]: why do you use name to get inputs above but placeholder text here? placeholder text is also more brittle

File: src/components/SignUpForm/index.test.js [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+      `${"a".repeat(MIN_CHAR - 1)}`
+    );
+    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
+      `${"a".repeat(MIN_CHAR - 1)}`
+    );
+    expect(screen.getByRole("checkbox").checked).toBe(true);
+  });
+
+  it("should display max length error when password is long", async () => {
+    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
```
Comment (louisrli) [inline src/components/SignUpForm/index.test.js]: not sure why this is in this test but if you are relying on the order of the fields for testing practice this is also not great structuring.


imagine that I move the first name field below the password field. if this breaks a lot of tests and I have to go through each one of your unit tests to fix them, then something needs to be changed either about the structure of the code (not sure if it's showing one error at a time right now)

basically as a reader I am reading a test about max length of password and I see a lot of stuff about first and last name and I'm confused

File: src/components/SignUpForm/index.test.js [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+    );
+    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
+    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
+      "password"
+    );
+    expect(screen.getByRole("checkbox").checked).toBe(false);
+  });
+
+  it("should not display error when value is valid", async () => {
+    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
```
Comment (louisrli) [inline src/components/SignUpForm/index.test.js]: in general I'd say these tests are too long and helper functions need to be used liberally for both readability and maintainability 

you should think about the ease of updating these tests if something changes. right now there's a lot of scenarios where a small change requires updating in like 10 places in the test file

**Review by louisrli — CHANGES_REQUESTED** [review]

File: src/components/SignUpForm/index.test.js (hunk 9 above)
Comment (jennetturkkan) [reply src/components/SignUpForm/index.test.js]: file name should be index.test.jsx for consistency @Jamil-Bailony

File: src/components/SignUpForm/__snapshots__/index.test.js.snap (hunk 5 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/__snapshots__/index.test.js.snap]: that's right, when I edit anything, there would be two snapshots edit

File: src/components/SignUpForm/index.jsx (hunk 6 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.jsx]: I put just for learning purpose :)
I will remove it

File: src/components/SignUpForm/index.jsx (hunk 7 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.jsx]: done!

File: src/components/SignUpForm/index.test.js (hunk 9 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.test.js]: we talked about it, the test file is a JavaScript file not a JSX file @jennethydyrova

Comment (jennetturkkan) [reply src/components/SignUpForm/index.test.js]: I don't remember that we talked about it but whole team is using jsx if you check components

Comment (jennetturkkan) [reply src/components/SignUpForm/index.test.js]: but I guess you used js in all your tests. we should change then in other components

Comment (louisrli) [reply src/components/SignUpForm/index.test.js]: actually its jsx. because you are using jsx

Comment (louisrli) [reply src/components/SignUpForm/index.test.js]:
```
    render(<SignUpForm submit={mockSubmit} />);
```

code like this is not allowed in normal js. it will also not highlight properly in .js. therefore its jsx

File: src/components/SignUpForm/index.test.js (hunk 10 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.test.js]: you are totally right, I fixed that, thank you very much

File: src/components/SignUpForm/index.test.js (hunk 11 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.test.js]: this is the more direct and easy way to reach elements as I found. Unfortunately, the [queries](https://testing-library.com/docs/dom-testing-library/api-queries/#queries) that React testing library use are so few and depend in general on the content shown.

File: src/components/SignUpForm/index.test.js (hunk 12 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.test.js]: First, filling the input fields does not relay on any order, so I will try to move the input field we are focusing on and testing to the top of the unite. Second, I have to insert a valid value into those filed other that the field that I am testing to check that there is just one error. For error messages shown, we are showing just one error message at a time.

File: src/components/SignUpForm/index.test.js (hunk 13 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.test.js]: When I wrote that code, I did think that it will be too long. I think this task is not so easy and need much time, so I added a TODO comment to work on it later if you do not mind.

File: src/components/SignUpForm/index.jsx (hunk 8 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.jsx]: I had used {' '} but `yarn fix` removed them later :/

**Comment by louisrli** [comment]
yarn run fix isn't going to replace {' '} with nbsp as far as i know.
either try again or show me exactly what you put before and after

**Comment by Jamil-Bailony** [comment]
> yarn run fix isn't going to replace {' '} with nbsp as far as i know

I did not mean that, I said that it replace `{" "}` with a typical space. Before running `yarn fix`:
image

After:
image

I had an alert from esLint about placing the `<a>` tag in a new line but yarn fix is stubborn sometimes :/ @louisrli

**Comment by louisrli** [comment]
the second picture is completely fine and should be kept that way

you want your spaces to be breaking

**Comment by Jamil-Bailony** [comment]
okay, then that's fine. is there anything else? I would appreciate it if you can approve this PR.

Thank you @louisrli

**Comment by louisrli** [comment]
I haven't reviewed it yet so.... no

**Comment by Jamil-Bailony** [comment]
take you time, do not worry :)

**Review by mohammad-jundi — APPROVED** [review]

File: src/components/SignUpForm/index.jsx [hunk 14]
```diff
@@ -1,5 +1,134 @@
...
+            <input
+              className={errors.acceptTerms && "inputError"}
+              type="checkbox"
+              id="acceptTerms"
+              name="acceptTerms"
+              value="accepted"
+              aria-label="acceptTerms"
+              ref={register()}
+            />
+            I have read and agree with the <a href="/terms">Terms of Service</a>{" "}
```
Comment (mohammad-jundi) [inline src/components/SignUpForm/index.jsx:123]: extra space at the end of the line

**Review by mohammad-jundi — APPROVED** [review]

File: src/components/InputErrorMessage/index.jsx [hunk 15]
```diff
@@ -2,11 +2,20 @@ import React from "react";
...
 
-export default function InputErrorMessage({ message }) {
-  return (
-    <div className="errorMessage" role="alert">
-      <img src={patchExclamationFill} alt="Error Logo" width="32" height="32" />
-      <p>{message}</p>
-    </div>
-  );
+export default function InputErrorMessage({ errors, name }) {
+  if (errors && errors[name]) {
```
Comment (louisrli) [inline src/components/InputErrorMessage/index.jsx]: avoid nesting

if (!errors || !errors[name]) {
return null;
}
return x

File: src/components/InputErrorMessage/index.jsx [hunk 16]
```diff
@@ -2,11 +2,20 @@ import React from "react";
...
+    return (
+      <div className="errorMessage" role="alert">
+        <img
+          src={patchExclamationFill}
+          alt="Error Logo"
+          width="32"
```
Comment (louisrli) [inline src/components/InputErrorMessage/index.jsx]: use css for width and height

File: src/components/InputErrorMessage/index.jsx [hunk 17]
```diff
@@ -2,11 +2,20 @@ import React from "react";
...
-      <p>{message}</p>
-    </div>
-  );
+export default function InputErrorMessage({ errors, name }) {
+  if (errors && errors[name]) {
+    return (
+      <div className="errorMessage" role="alert">
+        <img
+          src={patchExclamationFill}
+          alt="Error Logo"
```
Comment (louisrli) [inline src/components/InputErrorMessage/index.jsx]: Error logo

File: src/components/SignUpForm/index.jsx [hunk 18]
```diff
@@ -1,5 +1,134 @@
...
+  }),
+  confirmPassword: Joi.string()
+    .valid(Joi.ref("password"))
+    .required()
+    .strict()
+    .messages({
+      "any.only": "Those passwords didn't match",
+    }),
+  acceptTerms: Joi.valid("accepted").required().messages({
+    "any.only": `You should accept our terms of service and privcey policy`,
```
Comment (louisrli) [inline src/components/SignUpForm/index.jsx]: privacy

Comment (louisrli) [reply src/components/SignUpForm/index.jsx]: must accept

File: src/components/SignUpForm/index.jsx [hunk 19]
```diff
@@ -1,5 +1,134 @@
...
+              name="firstName"
+              aria-label="firstName"
+              ref={register()}
+            />
+            <InputErrorMessage errors={errors} name="firstName" />
+          </div>
+
+          <div className="formInput">
+            <input
+              className={errors.email && "inputError"}
```
Comment (louisrli) [inline src/components/SignUpForm/index.jsx]: spot the error; this is why you don't copy-paste code

make a function called
getErrorClass

its called as
getErrorClass(errors.email)

File: src/components/InputErrorMessage/__snapshots__/index.test.js.snap [hunk 20]
```diff
@@ -1,16 +1,3 @@
...
-  <img
-    alt="Error Logo"
-    height="32"
-    src="patchExclamationFill.svg"
-    width="32"
-  />
-  <p />
-</div>
-`;
+exports[`renders correctly and matches the snapshot 1`] = `null`;
```
Comment (louisrli) [inline src/components/InputErrorMessage/__snapshots__/index.test.js.snap]: ?

File: src/components/InputErrorMessage/index.jsx (hunk 15 above)
Comment (Jamil-Bailony) [reply src/components/InputErrorMessage/index.jsx]: okay, done!

File: src/components/SignUpForm/index.jsx (hunk 19 above)
Comment (Jamil-Bailony) [reply src/components/SignUpForm/index.jsx]: I see, your notes make me take a step back and re-read my code with Louis's eyes, thank you

**Comment by Jamil-Bailony** [comment]
if there isn't anything else, I was wondering if you can take a look at it when you can. Thank you @louisrli

File: src/components/InputErrorMessage/index.jsx [hunk 21]
```diff
@@ -2,11 +2,15 @@ import React from "react";
 import "./index.scss";
 import patchExclamationFill from "../../images/icons/patchExclamationFill.svg";
 
-export default function InputErrorMessage({ message }) {
+export default function InputErrorMessage({ error }) {
+  if (!error) {
```
Comment (louisrli) [inline src/components/InputErrorMessage/index.jsx:6]: shouldn't it check error.msg

File: src/helperFunctions.js [hunk 22]
```diff
@@ -0,0 +1,7 @@
+export function getErrorClass(error) {
```
Comment (louisrli) [inline src/helperFunctions.js]: please put this under components/errorhelpers or something. not in src.

File: src/components/SignUpForm/index.test.jsx [hunk 23]
```diff
@@ -0,0 +1,415 @@
+import React from "react";
+import { render, screen, fireEvent, waitFor } from "@testing-library/react";
+import "@testing-library/jest-dom/extend-expect";
+
+import SignUpForm from "./index";
+
+// TODO: refactor the code and extract repetitive code to helper functions
```
Comment (louisrli) [inline src/components/SignUpForm/index.test.jsx:7]: that comment wasn't for a separate PR. like I said we don't commit incomplete code because there is high probability no one will fix this PR

fix and identify the common functions, the rest file should not be 400 lines

**Review by louisrli — CHANGES_REQUESTED** [review]

File: src/components/InputErrorMessage/index.jsx (hunk 21 above)
Comment (Jamil-Bailony) [reply src/components/InputErrorMessage/index.jsx:6]: what did you mean? can you explain your point of view please?

This component will get the error object that related to each input validation. For example, I will render it for an email input like:
```
<InputErrorMessage error={errors.email} />
```
so if that error, related to email, not exists, the if part will handle it or the error message will be rendered and the message will be there for sure as the `errors.email` is there.

File: src/helperFunctions.js (hunk 22 above)
Comment (Jamil-Bailony) [reply src/helperFunctions.js]: I found one on reddit said that we can put our helpers in a folder called utils so I put it in that folder in src. I wish that is fine.

Comment (louisrli) [reply src/helperFunctions.js]: what are your teammates doing? most important thing is consistency in codebase, if no one else has helpers you can src/utils/errorHelpers

helperfunctions too general a name. adn dont pollute src/ top level

File: src/components/InputErrorMessage/index.jsx (hunk 21 above)
Comment (louisrli) [reply src/components/InputErrorMessage/index.jsx:6]: is error.message guaranteed to be populated?

**Review by louisrli — APPROVED** [review]
its fine just fix the filenames

File: src/components/InputErrorMessage/index.jsx (hunk 21 above)
Comment (Jamil-Bailony) [reply src/components/InputErrorMessage/index.jsx:6]: yeah it must, so how the component would know if there is an error, as I should not be rendered in normal cases where there is not any validation error.

Comment (louisrli) [reply src/components/InputErrorMessage/index.jsx:6]: i'm not sure i understand your answer but the safer thing to do here is to also add a check for error.message being undefined at the top. you shouldn't assume the fields are present
